### PR TITLE
feat: allow multi language spellcheckprovider

### PR DESCRIPTION
**Pull request by kwonoj**
#### Description of Change
- Relates to #15202, #7244

This PR attempts to expand `setSpellCheckProvider` to support multi-language spellchecker. Consumer of interface can opt-in by specifying multiple languages when initializing spellcheck provider. SpellcheckClient will create wordIterator per each language specified, will call provider function with object have language as keys, and word to check spell as its values.

Marking misspelling logic is not changed, as long as async callback returns word, it'll be marked as error. It's up to consumer's decision as same as before.
 
#### Checklist

- [ ] PR description included and stakeholders cc'd
- [ ] `npm test` passes
- [x] tests are [changed](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [ ] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [ ] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: `webFrame.setSpellcheckerProvider` supports multiple languages.

### CODE REVIEW

**Comment by kwonoj** [comment]
@nitsakh and I discussed that actually, maybe it's possible - but it requires languages param (`string` | `Array<string>`) and spellcheck fn's param (`Array<string> | Record<string, Array<string>`) both. 

> when you have to check Electron version in your code

Majority of case application relies on certain version of electron and when they update it they just need to apply breaking changes, isn't it? I agree breaking changes could require effort to migration, don't think many cases single code runs on multiple electron version.

**Comment by miniak** [comment]
@kwonoj in our case unfortunately due to the complicated way we deploy changes we need to be able to build the same code with multiple versions of Electron. We had to do a version check to handle the previous breaking change made in Electron 5 (which we backported to out internal Electron 4 builds to make it even more complicated).

**Comment by kwonoj** [comment]
OK. I personally think this would be ok to be breaking (especially migration should be easy enough) but don't have strong opinion. Will try to update changes to interop with current interfaces.

File: docs/api/web-frame.md [hunk 1]
```diff
@@ -65,6 +65,15 @@ Sets the maximum and minimum layout-based (i.e. non-visual) zoom level.
...
 ### `webFrame.setSpellCheckProvider(language, provider)`
 
+* `language` String[]
+* `provider` Object
+  * `spellCheck` Function
+    * `words` Record<String, String[]>
+    * `callback` Function
+      * `misspeltWords` String[]
+
+### `webFrame.setSpellCheckProvider(language, provider)`
```
Comment (kwonoj) [inline docs/api/web-frame.md]: Not sure what would be recommended way to create overload definition. Specifying union type (`language: String | String[]` and `words: String[] | Record<String, String[]>`) cannot narrow down param types, especially for `words` param `String[] | Record<String, String[]>` does not have interesection it's force required to cast types (which could be breaking changes in terms of types).

To make this work, listed each definition separately to create overloaded fn signatures.

**Comment by kwonoj** [comment]
@miniak @MarshallOfSound PR updated to support existing interfaces.

File: shell/renderer/api/atom_api_web_frame.cc [hunk 2]
```diff
@@ -314,6 +314,16 @@ void SetSpellCheckProvider(mate::Arguments* args,
     return;
   }
 
+  std::vector<std::string> language_vector;
+  bool is_multi_lang_spell_check = languages->IsArray();
+  if (is_multi_lang_spell_check) {
+    gin::ConvertFromV8(args->isolate(), languages, &language_vector);
```
Comment (MarshallOfSound) [inline shell/renderer/api/atom_api_web_frame.cc]: This conversion could fail, it should be handled

File: shell/renderer/api/atom_api_web_frame.cc [hunk 3]
```diff
@@ -314,6 +314,16 @@ void SetSpellCheckProvider(mate::Arguments* args,
...
+  } else {
+    std::string language;
+    gin::ConvertFromV8(args->isolate(), languages, &language);
```
Comment (MarshallOfSound) [inline shell/renderer/api/atom_api_web_frame.cc]: Same with this conversion, if both fail we should throw an error

**Comment by miniak** [comment]
Is this still relevant after #20692?

**Comment by zcbenz** [comment]
I'm closing this PR since with #20692 this PR should be unnecessary. Please let me know if I'm wrong and I will reopen this.

**Pull request closed without merging**